src/runtime/ndarray.cc

```diff
@@ -256,16 +257,26 @@ void NDArray::CopyFromTo(DLTensor* from,
     from_size, from->ctx, to->ctx, from->dtype, stream);
 }
 
-void NDArray::PinData(DLTensor* tensor) {
-  if (IsDataPinned(tensor)) return;
+void NDArray::PinContainer(NDArray::Container* ptr) {
+  if (IsContainerPinned(ptr)) return;
+  auto* tensor = &(ptr->dl_tensor);
   CHECK_EQ(tensor->ctx.device_type, kDLCPU)
     << "Only NDArray on CPU can be pinned";
   DeviceAPI::Get(kDLGPU)->PinData(tensor->data, GetDataSize(*tensor));
+  ptr->pinned_by_dgl_ = true;
 }
 
-void NDArray::UnpinData(DLTensor* tensor) {
-  if (!IsDataPinned(tensor)) return;
-  DeviceAPI::Get(kDLGPU)->UnpinData(tensor->data);
+void NDArray::UnpinContainer(NDArray::Container* ptr) {
+  auto container_is_pinned = IsContainerPinned(ptr);
+  // The tensor may be pinned outside of DGL via a different CUDA API,
+  // so we cannot unpin it with cudaHostUnregister.
+  CHECK(ptr->pinned_by_dgl_ || !container_is_pinned)
+    << "Cannot unpin a tensor that is pinned outside of DGL.";
```

Comment:
Just realized this case and I'd like to give it a clearer error message here. There is also a unit test for it:
https://github.com/dmlc/dgl/blob/d53adc16f0ac43cfd503cddb4bb1c7bff14dbb38/tests/compute/test_pin_memory.py#L22-L25.

However, this would break the previous `unpin_memory_inplace`:
```python
F.zerocopy_to_dgl_ndarray_for_write(tensor).unpin_memory_()
```
because the temporary ndarray created here doesn't have the `pinned_by_dgl_` flag. 

I suggest remove `unpin_memory_inplace()` and unpin the array via `nd.unpin_memory_()` where `nd` is returned from `pin_memory_inplace()`.